exec/java-exec/src/main/java/org/apache/drill/exec/record/VectorContainer.java

```diff
@@ -210,6 +214,109 @@ public void addCollection(Iterable<ValueVector> vectors) {
     }
   }
 
+  /**
+    * This works with non-hyper {@link VectorContainer}s which have no selection vectors.
+    * Appends a row taken from a source {@link VectorContainer} to this {@link VectorContainer}.
+    * @param srcContainer The {@link VectorContainer} to copy a row from.
+    * @param srcIndex The index of the row to copy from the source {@link VectorContainer}.
+    * @return Position where the row was appended
+    */
+    public int appendRow(VectorContainer srcContainer, int srcIndex) {
+      for (int vectorIndex = 0; vectorIndex < wrappers.size(); vectorIndex++) {
+        ValueVector destVector = wrappers.get(vectorIndex).getValueVector();
+        ValueVector srcVector = srcContainer.wrappers.get(vectorIndex).getValueVector();
+        destVector.copyEntry(recordCount, srcVector, srcIndex);
+      }
+      int pos = recordCount++;
+      initialized = true;
+      return pos;
+    }
+
+  /**
+   *  This method currently is only used by the Hash Join to return a row composed of build+probe rows
+   *
+   * This works with non-hyper {@link VectorContainer}s which have no selection vectors.
+   * Appends a row taken from two source {@link VectorContainer}s to this {@link VectorContainer}.
+   * @param buildSrcContainer The {@link VectorContainer} to copy the first columns of a row from.
+   * @param buildSrcIndex The index of the row to copy from the build side source {@link VectorContainer}.
+   * @param probeSrcContainer The {@link VectorContainer} to copy the last columns of a row from.
+   * @param probeSrcIndex The index of the row to copy from the probe side source {@link VectorContainer}.
+   * @return Number of records in the container after appending
+   */
+  public int appendRowXXX(VectorContainer buildSrcContainer, int buildSrcIndex, VectorContainer probeSrcContainer, int probeSrcIndex) {
+    if ( buildSrcContainer != null ) {
+      for (int vectorIndex = 0; vectorIndex < buildSrcContainer.wrappers.size(); vectorIndex++) {
+        ValueVector destVector = wrappers.get(vectorIndex).getValueVector();
+        ValueVector srcVector = buildSrcContainer.wrappers.get(vectorIndex).getValueVector();
+        destVector.copyEntry(recordCount, srcVector, buildSrcIndex);
+      }
+    }
+    if ( probeSrcContainer != null ) {
+      int baseIndex = wrappers.size() - probeSrcContainer.wrappers.size();
+      for (int vectorIndex = baseIndex; vectorIndex < wrappers.size(); vectorIndex++) {
+        ValueVector destVector = wrappers.get(vectorIndex).getValueVector();
+        ValueVector srcVector = probeSrcContainer.wrappers.get(vectorIndex).getValueVector();
+        destVector.copyEntry(recordCount, srcVector, probeSrcIndex);
+      }
+    }
+    recordCount++;
+    initialized = true;
+    return recordCount;
+  }
+
+  private int appendBuild(VectorContainer buildSrcContainer, int buildSrcIndex) {
```

Comment:
Specific to HashJoin, move out of VectorContainer.

Reply:
@Ben-Zvi In our discussion offline you were concerned about the overhead of moving these methods out of VectorContainer. Specifically you were concerned about the overhead of calling a get Method. People have benchmarked this overhead and analyzed the assembly generated by the JIT. It turns out calling a getter method generates the same assembly code as a local field access. Please see this post https://stackoverflow.com/questions/23931546/java-getter-and-setter-faster-than-direct-access . So there is no performance difference.

Since there is no performance benefit to adding HashJoin code to VectorContainer, please move these methods out. If all the operators start adding code to VectorContainer the class will become unwieldy.

Reply:
Done ....